From 1aa976db79e094d148110667231c7a75ffa7e1d5 Mon Sep 17 00:00:00 2001
From: ingmar-k <ingmar.klein@hs-augsburg.de>
Date: Sat, 9 Aug 2014 19:47:15 +0200
Subject: [PATCH 3/4] Fix small Kirkwood cpufreq error.

---
 drivers/cpufreq/Kconfig.arm | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/cpufreq/Kconfig.arm b/drivers/cpufreq/Kconfig.arm
index 5805035..2d1715c 100644
--- a/drivers/cpufreq/Kconfig.arm
+++ b/drivers/cpufreq/Kconfig.arm
@@ -118,7 +118,7 @@ config ARM_INTEGRATOR
 	  If in doubt, say Y.
 
 config ARM_KIRKWOOD_CPUFREQ
-	def_bool MACH_KIRKWOOD
+	def_bool ARCH_KIRKWOOD || MACH_KIRKWOOD
 	help
 	  This adds the CPUFreq driver for Marvell Kirkwood
 	  SoCs.
-- 
2.1.0.rc1

